File: grundlagen_der_regelungstechnik/src/steuerbarkeit.m
function check_controllability(A, b)
    % CHECK_CONTROLLABILITY: Check if the system (A, b) is controllable.
    %
    % Inputs:
    %   A - State matrix (n x n)
    %   b - Input matrix (n x 1)
    %
    % Example usage:
    %   A = [1, 2; 0, -1];
    %   b = [0; 1];
    %   check_controllability(A, b);

    % Get the size of the state matrix A
    n = size(A, 1);
    
    % Initialize the controllability matrix
    ControllabilityMatrix = [];
    
    % Build the controllability matrix
    for i = 0:n-1
        ControllabilityMatrix = [ControllabilityMatrix, A^i * b];
    end
    
    % Calculate the rank of the controllability matrix
    rank_C = rank(ControllabilityMatrix);
    
    % Check controllability
    if rank_C == n
        fprintf('The system is controllable (rank = %d).\n', rank_C);
    else
        fprintf('The system is NOT controllable (rank = %d).\n', rank_C);
    end
end

File: grundlagen_der_regelungstechnik/src/steuerbarkeit.py
import numpy as np

def is_controllable(A, b):
    """
    Check if the system (A, b) is controllable.

    Parameters:
        A (ndarray): State matrix (n x n)
        b (ndarray): Input matrix (n x 1)

    Returns:
        bool: True if the system is controllable, False otherwise.
    """
    # Number of states (n)
    n = A.shape[0]

    # Construct the controllability matrix
    steuermatrix = b
    for i in range(1, n):
        steuermatrix = np.hstack((steuermatrix, np.linalg.matrix_power(A, i) @ b))

    # Check if the controllability matrix is full rank
    rank = np.linalg.matrix_rank(steuermatrix)
    
    return rank == n

# Example usage
A = np.array([[1, 1], [0, -1]])  # Example A matrix
b = np.array([[0], [1]])         # Example b vector

if is_controllable(A, b):
    print("The system is controllable.")
else:
    print("The system is NOT controllable.")